pkg/provisioner/controllerprovisioner_azuredisk.go

```diff
@@ -0,0 +1,271 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provisioner
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"strconv"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2020-12-01/compute"
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	azDiskClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/azureutils"
+	azure "sigs.k8s.io/cloud-provider-azure/pkg/provider"
+	"sigs.k8s.io/cloud-provider-azure/pkg/retry"
+)
+
+type ControllerProvisioner struct {
```

Comment:
There should be a `fakeControllerProvisioner` to support server component unit tests. Ideally, we would have a fake Azure cloud client. We could either stand-up a fake here or use [`gomock`](https://pkg.go.dev/github.com/golang/mock/gomock). @sunpa93 has some experience with `gomock`.

Reply:
If there is an interface that `ControllerProvisioner` currently implements, you can `mock-gen -package=<mock package name> -destination=<output path> -source=<source file where the interface is defined>` to generate a mock interface for testing. Otherwise, you can define an interface for the ControllerProvisioner to implement and run mock-gen on it.

(Not entirely sure about this but) it seems to be a convention to name the interface `Interface` and also output generate mock interface to `interface.go` under a separate mock folder along with `doc.go` file.

You can use the generate mock interface in test by importing it and initializing it with `NewMockInterface(ctrl *gomock.Controller)`.  From that point, you can simulate any function call defined in mock interface with specific input and outputs by calling `<Initialized Mock Interface>.EXPECT().<Function Name>(<Function Paramters>...).Return(<Desired Return Value>)`. You can also make the calls return results in specific orders and much more.

I am no expert but let me know if things are unclear :)